### PR TITLE
Rdm 13132 - FTAs for CategoryID columns in ComplexTypes tab

### PR DESCRIPTION
**Pull request by sullivangareth**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-13132

### Change description ###

FTAs for CategoryID columns in ComplexTypes tab

This PR is based on PR for ComplexID in CaseField tab 
https://github.com/hmcts/ccd-definition-store-api/pull/1171)

It is also dependant on a PR for release of ccd-test-definitions 
https://github.com/hmcts/ccd-test-definitions/pull/43

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -266,7 +266,7 @@ subprojects { subproject ->
         testCompile "io.rest-assured:rest-assured:${restAssuredVersion}"
         testCompile group: 'org.powermock', name: 'powermock-api-mockito2', version: powermockVersion
         testCompile group: 'org.powermock', name: 'powermock-module-junit4', version: powermockVersion
-        testCompile group: 'com.github.hmcts', name: 'ccd-test-definitions', version: '7.14.1_RDM-13132_2'
+        testCompile group: 'com.github.hmcts', name: 'ccd-test-definitions', version: '7.14.1_RDM-13132_3'
```
Comment (mattnayler) [inline build.gradle:269]: This needs to be updated to a full release version of the hmcts/ccd-test-defintions before going into master. see https://github.com/hmcts/ccd-test-definitions/pull/43.

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 2]
```diff
@@ -263,4 +263,43 @@ public void testRegisterCaseTypeForCategoryWithSameIdTwiceShouldFail() {
...
+        FieldTypeEntity fieldTypeEntity2 = new FieldTypeEntity();
+        fieldTypeEntity2.setReference("Entity2");
+        FieldTypeEntity notComplexEntity = new FieldTypeEntity();
+        notComplexEntity.setReference("NotComplex");
+        fieldTypeEntity2.setBaseFieldType(notComplexEntity);
+
+        parseContext.addToAllTypes(fieldTypeEntity1);
+        parseContext.addToAllTypes(fieldTypeEntity2);
+
+        assertEquals(1, parseContext.getComplexTypes().size());
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:283]: https://github.com/hmcts/ccd-definition-store-api/pull/1177/commits/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 3]
```diff
@@ -263,4 +263,43 @@ public void testRegisterCaseTypeForCategoryWithSameIdTwiceShouldFail() {
...
+        CategoryEntity categoryEntity = new CategoryEntity();
+        categoryEntity.setCategoryId("A");
+        String caseTypeId = "CaseTypeId";
+        parseContext.registerCaseTypeForCategory(caseTypeId, categoryEntity);
+        Assertions.assertTrue(parseContext.checkCategoryExists("A"));
+    }
+
+    @Test
+    public void testIsCheckCategoryDoesNotExists() {
+        Assertions.assertFalse(parseContext.checkCategoryExists("A"));
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java [hunk 4]
```diff
@@ -59,9 +61,135 @@ public CaseFieldEntity buildCaseTypeDocumentReference(String reference, String c
...
+
+    @Test
+    public void testValidateCaseFieldFieldTypeCollectionOfDocument() {
+        parseContext = buildParseContextCaseTypeEntity(COLLECTION, DOCUMENT, DIVORCE_DOCS);
+
+        CategoryEntity category1 = new CategoryEntity();
+        category1.setCategoryId(DIVORCE_DOCS);
+        parseContext.registerCaseTypeForCategory(CASE_TYPE, category1);
+        categoryValidator.validate(parseContext);
+    }
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java [hunk 5]
```diff
@@ -59,9 +61,135 @@ public CaseFieldEntity buildCaseTypeDocumentReference(String reference, String c
...
+
+    @Test
+    public void testValidateCaseFieldFieldTypeCollectionOfDocumentNullCategory() {
+        parseContext = buildParseContextCaseTypeEntity(COLLECTION, DOCUMENT, null);
+
+        CategoryEntity category1 = new CategoryEntity();
+        category1.setCategoryId(DIVORCE_DOCS);
+        parseContext.registerCaseTypeForCategory(CASE_TYPE, category1);
+        categoryValidator.validate(parseContext);
+    }
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.13.td.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{
+  "_guid_": "S-106.13",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.13.td.json]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.14.td.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
+{
+  "_guid_": "S-106.14",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.14.td.json]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.15.td.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{
+  "_guid_": "S-106.15",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.15.td.json]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.16.td.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
+{
+  "_guid_": "S-106.16",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.16.td.json]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.17.td.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
+{
+  "_guid_": "S-106.17",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.17.td.json]: Fixed in https://github.com/hmcts/ccd-definition-store-api/commit/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/F-106.feature [hunk 11]
```diff
@@ -26,7 +26,8 @@ Feature: F-106 New CategoryId Column
 
   @S-106.3
   @AC-3 @AC-4 @AC-7 @AC-8
-  Scenario: Import definition file with valid CategoryID column in CaseField tab – should return 200 response
+  @AC-15 @AC-16 @AC-19 @AC-20
+  Scenario: Import definition file with valid CategoryID column in CaseField and ComplexType tabs – should return 200 response
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/F-106.feature:30]: Fixed in https://github.com/hmcts/ccd-definition-store-api/pull/1177/commits/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.12.td.json [hunk 12]
```diff
@@ -0,0 +1,31 @@
+{
+  "_guid_": "S-106.12",
+  "_extends_": "Import_Base_data",
+  "title": "Import Definition file without the CategoryID column in the ComplexTypes tab",
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply aat/src/aat/resources/features/F-106 New CategoryID Column/S-106.12.td.json]: fixed in https://github.com/hmcts/ccd-definition-store-api/pull/1177/commits/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ComplexFieldParserTest.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class ComplexFieldParserTest extends ParserTestBase {
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ComplexFieldParserTest.java]: Fixed in https://github.com/hmcts/ccd-definition-store-api/pull/1177/commits/ddd7d2a24fd6c44a2e3b3ce05162ca702f817286

**Review by mattnayler — APPROVED** [review]
approved: subject to the minor change raised against `build.gradle`.

**Pull request merged**